src/job.rs

```diff
@@ -1,16 +1,102 @@
     pub environment: HashMap<String, String>,
+    pub working_directory: PathBuf,
+    pub inputs: Vec<PathBuf>,
+    pub outputs: Vec<PathBuf>,
 }
 
 impl Job {
-    pub fn run(&self) -> io::Result<Output> {
+    pub fn run(&self) -> Result<Output> {
+        let work_dir = Builder::new()
+            .prefix(format!("job-{}", self.command).as_str())
+            .tempdir()
+            .context("couldn't create a temporary directory for the job")?;
+
+        self.prepare_workspace(work_dir.path())
+            .context("couldn't prepare files in the working directory")?;
+
+        let output = self
+            .run_job(work_dir.path())
+            .context("couldn't run the job")?;
+
+        self.tear_down_workspace(work_dir.path())
+            .context("couldn't tear down the workspace")?;
+
+        work_dir
+            .close()
+            .context("couldn't clean up the job's temporary directory")?;
+
+        Ok(output)
+    }
+
+    fn prepare_workspace(&self, work_dir: &Path) -> Result<()> {
+        for input in &self.inputs {
+            let meta = fs::metadata(input)
+                .with_context(|| format!("couldn't read metadata for {}", input.display()))?;
+
+            // Copy all the files over. Note that we can't just create symlinks
+            // to the files in all cases because that would allow the job code
+            // to modify source files on disk as a side effect. We probably want
+            // to allow at least directory symlinks in the future (for caches,
+            // for instance) and we'll need to extend the `inputs` concept then.
+            if meta.is_dir() {
+                for item in WalkDir::new(input) {
+                    let entry =
+                        item.with_context(|| format!("couldn't walk through {}", input.display()))?;
+
+                    let dest = self
+                        .path_in_workspace(work_dir, &entry.path().to_path_buf())
+                        .with_context(|| {
+                            format!(
+                                "couldn't get a path to {} in the workspace",
+                                input.display()
+                            )
+                        })?;
+
+                    if entry.file_type().is_dir() {
+                        fs::create_dir(&dest)
+                            .with_context(|| format!("couldn't create {}", dest.display()))?
+                    } else {
+                        // assuming file or symlink
+                        fs::copy(entry.path(), &dest).with_context(|| {
+                            format!(
+                                "couldn't copy {} to {}",
+                                entry.path().display(),
+                                dest.display()
+                            )
+                        })?;
+                    }
+                }
+            } else {
+                // it's a file, or maybe a symlink
+                self.copy_creating_directories(
```

Comment:
I could be thinking about this wrong, but shouldn't we be able to do a normal `fs::copy` here? In other words, shouldn't the necessary destination directories already have been created in the workspace by the time we traverse to any files that would go in them?

Reply:
Added a test to demonstrate that this is necessary in 5583369